### PR TITLE
Add syscall/js

### LINKED ISSUE #899: Unify GopherJS's `syscall/js` and GopherWasm
@neelance suggested to unify [GopherWasm](https://github.com/gopherjs/gopherwasm) and GopherJS by making GopherJS recognize `syscall/js` as a wrapper like GopherWasm. This would enable the users to replace

```
import "github.com/gopherjs/gopherwasm/js"
```

with

```
import "syscall/js"
```

while the programs still work both for GopherJS and Wasm. This simplifies things very much and encourages users to migrate from GopherJS to Wasm more smoothly than the current GopherWasm.

Any thoughts?

### PR DESCRIPTION
**Pull request by hajimehoshi**
This PR adds a new package `syscall/js`. Originally `syscall/js` is for WebAssembly, but in GopherJS this package works as a wrapper of `github.com/gopherjs/gopherjs/js`. With this package, developers can use `syscall/js` both for GopherJS and WebAssembly.

The implementation is basically same as `github.com/gohperjs/gopherwasm`.

Fixes #899

### CODE REVIEW

**Comment by hajimehoshi** [comment]
```
--- FAIL: TestSleep (0.10s)
    test.148654449:38: Sleep(100ms) slept for only 99ms
```

Meh, this sounds like a flaky issue

**Comment by neelance** [comment]
@hajimehoshi While working on WebAssembly I've noticed that `setTimeout` sometimes invokes the callback 1ms early. Maybe this is the same issue. It would probably be okay to always add 1ms to the delay.

**Comment by hajimehoshi** [comment]
@neelance Thank. I think I'll do this in another PR later.

File: compiler/natives/fs_vfsdata.go [hunk 1]
```diff
@@ -387,10 +387,10 @@ var FS = func() http.FileSystem {
...
...\xd3\xf8\xc1\x6c\x30\x0f\x5e\x11\xfe\x62\x6e\x62\xd6\x00\xf4\xc3\x87\x08\x86\x79\xf0\x62\x1e\x42\xf1\xeb\x5b\x4c\x78\x1a\xb9\xa0\x00\x5d\x9d\xf2\xeb\x34\xa4\xd4\x8b\xf9\xe9\xb5\x4f\x0d\x98\x71\x69\x38\x27\x1b\x52\x71\x51\xa2\x7f\xa3\x67\x3d\x7f\x7c\xd6\x76\x4e\x95\xcf\xb1\xff\xfa\x2f\xfd\x58\xcf\x55\x5f\xb6\x1c\xcc\x3b\x98\xf5\x60\x46\xff\x44\x3f\x2b\x9e\xd3\x8b\xf9\xbe\x59\xf9\x87\xc3\x6f\x7b\x2d\x05\x5b\x4c\x7e\xff\xac\xe1\xc0\x01\xf4\x77\x02\x26\x9e\xe0\x08\xa9\x67\xf7\x99\xa9\x07\x0b\x65\x36\x1b\x31\x77\xf4\xfe\x1e\x99\x3b\x8f\xd9\xcf\xd6\xa7\x32\x56\x8c\xbd\xec\xe3\x80\x01\x19\x65\x37\x20\x23\x29\x65\x5e\x33\x91\xfa\xb5\x1e\xce\x7e\x01\xa0\x7b\xec\x17\xdf\xcc\xd6\xd6\xe1\x68\xec\x64\x68\x56\x8c\x24\xf3\x7c\x23\x63\x3a\x99\xd0\xc3\x4a\xfb\x37\xd3\xda\xbf\x6e\x53\xfe\x95\x7a\x9b\xba\x7b\xa7\xed\x46\xf8\x44\xbd\x4d\x83\xf3\x70\x87\x35\xf7\xd8\xde\xfa\xb0\xd7\xe9\x39\x88\x26\xea\xee\x41\x51\xe3\x98\xef\x16\x66\xd1\xfa\x28\x32\x8a\xee\xfb\xb8\xcc\x61\x8d\xe4\x21\x99\x33\x76\xbb\xb9\x00\xe3\x80\xc4\xef\x91\x4f\x23\x8d\x91\xfb\xf4\xb8\x60\x72\xf2\xc2\xcd\xc6\x44\x85\x4d\x30\x02\xc5\xb6\x0f\x03\xcc\xff\x96\xd6\x7f\x0d\x69\xb5\x65\x8f\x3d\x9e\x6c\x7f\x0e\x8e\x9f\xb2\x37\x02\xb5\x32\xcc
```
[the remark this answers is not in the document]
Comment (hajimehoshi) [reply compiler/natives/fs_vfsdata.go]: Thank you for triaging!

File: tests/syscalljs_test.go [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+package tests
+
+import (
+	"fmt"
+	"math"
+	"syscall/js"
+	"testing"
+)
+
+func TestSyscallJsNull(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (hajimehoshi) [reply tests/syscalljs_test.go]: Done.

**Comment by hajimehoshi** [comment]
@neelance Hi, are you happy with this PR? Thanks!

**Comment by hajimehoshi** [comment]
ping @neelance

**Comment by neelance** [comment]
Is this passing the normal tests of the `syscall/js` package?https://github.com/golang/go/blob/master/src/syscall/js/js_test.go

Is there any particular need for the custom test cases of this PR?

**Comment by hajimehoshi** [comment]
To test `js_test.go`, we'd need to copy the entire file except for `// +build js,wasm`. Should we do that?

**Comment by hajimehoshi** [comment]
BTW, I put a test file at `compiler/natives/src/syscall/js`, but the test was not executed, I'm not sure the reason though.

**Comment by neelance** [comment]
Shouldn't it be possible to modify `importWithSrcDir` to manually include `js_test.go` in the build?

**Comment by hajimehoshi** [comment]
Almost done! I found a problem in `js_test.go`:

```
func TestIntConversion(t *testing.T) {                                                                                                                                                                                                    
        testIntConversion(t, 0)                                                                                                                                                                                                           
        testIntConversion(t, 1)                                                                                                                                                                                                           
        testIntConversion(t, -1)                                                                                                                                                                                                          
        testIntConversion(t, 1<<20)                                                                                                                                                                                                       
        testIntConversion(t, -1<<20)                                                                                                                                                                                                      
        testIntConversion(t, 1<<40)                                                                                                                                                                                                       
        testIntConversion(t, -1<<40)                                                                                                                                                                                                      
        testIntConversion(t, 1<<60)                                                                                                                                                                                                       
        testIntConversion(t, -1<<60)                                                                                                                                                                                                      
}
```

This cannot be compiled on 32bit-integer environment and GopherJS is an environment with 32bit integers. Replacing the test function in a usual GopherJS way didn't work. Probably the const check happens before the replacing. Any suggestions?

**Comment by neelance** [comment]
This is unfortunate. Okay, maybe coping `js_test.go` is the simplest solution.

**Comment by hajimehoshi** [comment]
OK, I copied the test, but now I'm facing a new error on my machine:

```
$ go generate ./... && go run . test syscall/js
writing fs_vfsdata.go
writing fs_vfsdata.go
gopherjs: Source maps disabled. Install source-map-support module for nice stack traces. See https://github.com/gopherjs/gopherjs#gopherjs-run-gopherjs-test.
testing: warning: no tests to run
PASS
[...]
```

Looks like the file is compiled Now it's midnight, I'll investigate this further later...

**Comment by hajimehoshi** [comment]
@neelance In the current implementation, is it possible to 'add' new tests in compiler/native?

**Comment by hajimehoshi** [comment]
>  is it possible to 'add' new tests in compiler/native?

I thought it impossible, so I moved `js_test.go` in compiler/native to `tests/syscalljs/js_test.go`. Also I found some tests that cannot pass, so I added `t.Skip`.

* `TestNaN`: comparing `NaN` doesn't work with `*js.Object`
* `TestZeroValue`: the initial value of `*js.Object` is nil, which is not `undefined`.

Please take a look

**Comment by neelance** [comment]
The zero value is quite important. I think it should be possible to handle this correctly, shouldn't it?

**Comment by hajimehoshi** [comment]
> I think it should be possible to handle this correctly, shouldn't it?

Yes. I fixed that so that the zero value is js.Undefined.

I appreciate your patience and your support. Please take another look.

**Comment by hajimehoshi** [comment]
ping

**Comment by johanbrandhorst** [comment]
This is fantastic :tada:. Well done @hajimehoshi

**Pull request merged**